internal/controller/pdb_controller.go

```diff
@@ -0,0 +1,131 @@
+package controllers
+
+import (
+	"context"
+
+	"k8s.io/api/apps/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// DeploymentPDBReconciler reconciles a Deployment object and ensures an associated PDB is created and deleted
+type DeploymentPDBReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// Reconcile watches for Deployment changes (created, updated, deleted) and creates or deletes the associated PDB.
+func (r *DeploymentPDBReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Fetch the Deployment instance
+	var deployment v1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &deployment); err != nil {
+		log.Error(err, "unable to fetch Deployment")
+		return ctrl.Result{}, client.IgnoreNotFound(err) // Ignore if the deployment is not found (i.e., deleted)
+	}
+
+	// If the Deployment is deleted, ensure the associated PDB is also deleted
+	if deployment.DeletionTimestamp != nil {
+		return r.handleDeploymentDeletion(ctx, &deployment)
+	}
+
+	// If the Deployment is created or updated, ensure a PDB exists
+	return r.handleDeploymentCreation(ctx, &deployment)
+}
+
+// handleDeploymentCreation creates a PodDisruptionBudget when a Deployment is created or updated.
+func (r *DeploymentPDBReconciler) handleDeploymentCreation(ctx context.Context, deployment *v1.Deployment) (reconcile.Result, error) {
+	log := log.FromContext(ctx)
+	// Check if PDB already exists for this Deployment
+	pdb := &policyv1.PodDisruptionBudget{}
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: deployment.Namespace,
+		Name:      deployment.Name + "-pdb",
+	}, pdb)
+
+	if err == nil {
+		// PDB already exists, nothing to do
```

Comment:
So a fun todo for later is can we distinguish scales from our pdbwatcher from scales from other owners and keep minavailable up near replicas. Like if I start a deployment at 3 but then later say this is popular let me bump it to 5 should our pdb change.